proposals/extensions_v2.md

```diff
@@ -0,0 +1,375 @@
+
+```c#
+extension IListExtensions
+{
+    public void ForEach<T>(Action<T> act) for IList<T> list
+    {
+        foreach (var value in list)
+            act(list);
+    }
+
+    public long LongCount<T> for IList<T> list
+    {
+        get
+        {
+            long count = 0;
+            foreach (var value in list)
+                count++;
+
+            return count;
+        }
+    }
+}
+```
+
+Ideally with the optional first expansion we could 'lift' `List<T>` up to `extension IListExtensions`.  However, this doesn't work as we need to define the type parameter it references.  This naturally leads to the following idea:
+
+```c#
+extension IListExtensions<T> for IList<T> list
+{
+    public void ForEach(Action<T> act) { ... }
+    public long LongCount { ... }
+}
+```
+
+This has a few new, but solvable, design challenges.  For example, say one has the code:
+
+```c#
+List<int> ints = ...;
+var v = ints.ForEach(i => Console.WriteLine(i));
+```
+
+This naturally raises the question of how does this extension get picked for this particular receiver, and how does its type parameter get instantiated to the `int` type.
+
+Conceptually (and trying to keep somewhat in line with classic extension methods), we really want to think of the 'receiver' as an 'argument' to some method where normal type inference occurs.  Morally, we could think of there being a `IListExtension<T> Infer<T>(IList<T> list)` function whose shape is determined by the extension and its type-parameters and the extended receiver parameter.
+
+Then, when trying to determine if an extension applies to a receiver, it would be akin to calling that function with the receiver and seeing if inference works.  In the above example that would mean performing type inference on `Infer(ints)` seeing that `T` then bound to `int`, which then gives you back `IListExtensions<int>`.  At that point, lookup would then find and perform overload resolution on `ForEach(Action<int>)` with the lambda parameter.
+
+This approach does fundamentally take expand on the initial extension-members approach as now calling extensions is done in two-phases.  An initial phase to determine and infer extension type parameters based on the receiver, and  a second phase to determine and perform overload resolution on the member.
+
+We believe this is very powerful and beneficial.  But there are deep design questions here which may cause this to be scheduled after the core extension members work happens.
+
+## Expansion 4: Extensions as actual types
+
+We are very undecided on if we actually want this.  Currently, our view is that it feels like 'roles' fits this goal much better, especially if roles have the ability to be 'implicit' or 'explicit'.  Extensions exist very much in the space where they are erased and really are just delicious delicious sugar over calling effectively static helpers to augment a type or value.
```

Comment:
> especially if roles have the ability to be 'implicit' or 'explicit'.

I'm a little confused at how a role could be implicit without, unavoidably, providing extension members in a new way, unless implicit just means you can pass values without casting but lookup is unaffected. With the recent LDM decision to define roles as "explicit extensions," saying "implicit role" in this new context spins my head a bit.